modular_zubbers/code/modules/storyteller/storyteller.dm

```diff
@@ -0,0 +1,140 @@
+	var/list/point_gains_multipliers = list(
+		EVENT_TRACK_MUNDANE = 1,
+		EVENT_TRACK_MODERATE = 1,
+		EVENT_TRACK_MAJOR = 1,
+		EVENT_TRACK_ROLESET = 1,
+		EVENT_TRACK_OBJECTIVES = 1
+		)
+	/// Multipliers of weight to apply for each tag of an event.
+	var/list/tag_multipliers
+
+	/// Variance in cost of the purchased events. Effectively affects frequency of events
+	var/cost_variance = 15
+
+	/// Variance in the budget of roundstart points.
+	var/roundstart_points_variance = 15
+
+	/// Whether the storyteller guaranteed a roleset roll (antag) on roundstart. (Still needs to pass pop check)
+	var/guarantees_roundstart_roleset = TRUE
+
+	/// Whether the storyteller has the distributions disabled. Important for ghost storytellers
+	var/disable_distribution = FALSE
+
+	/// Whether people can vote for the storyteller
+	var/votable = TRUE
+	/// If defined, will need a minimum of population to be votable
+	var/population_min
+	/// If defined, it will not be votable if exceeding the population
+	var/population_max
+
+/datum/storyteller/process(delta_time)
+	if(disable_distribution)
+		return
+	add_points(delta_time)
+	handle_tracks()
+
+/// Add points to all tracks while respecting the multipliers.
+/datum/storyteller/proc/add_points(delta_time)
+	var/datum/controller/subsystem/gamemode/mode = SSgamemode
+	var/base_point = EVENT_POINT_GAINED_PER_SECOND * delta_time * mode.event_frequency_multiplier
+	for(var/track in mode.event_track_points)
+		var/point_gain = base_point * point_gains_multipliers[track] * mode.point_gain_multipliers[track]
+		if(mode.allow_pop_scaling)
+			point_gain *= mode.current_pop_scale_multipliers[track]
+		mode.event_track_points[track] += point_gain
+		mode.last_point_gains[track] = point_gain
+
+/// Goes through every track of the gamemode and checks if it passes a threshold to buy an event, if does, buys one.
+/datum/storyteller/proc/handle_tracks()
+	. = FALSE //Has return value for the roundstart loop
+	var/datum/controller/subsystem/gamemode/mode = SSgamemode
+	for(var/track in mode.event_track_points)
+		var/points = mode.event_track_points[track]
+		if(points >= mode.point_thresholds[track] && find_and_buy_event_from_track(track))
+			. = TRUE
+
+/// Find and buy a valid event from a track.
+/datum/storyteller/proc/find_and_buy_event_from_track(track)
+	. = FALSE
+	var/datum/controller/subsystem/gamemode/mode = SSgamemode
+	var/datum/round_event_control/picked_event
+	if(mode.forced_next_events[track]) //Forced event by admin
+		/// Dont check any prerequisites, it has been forced by an admin
+		picked_event = mode.forced_next_events[track]
+		mode.forced_next_events -= track
+	else
+		mode.update_crew_infos()
+		var/pop_required = mode.min_pop_thresholds[track]
+		if(mode.active_players < pop_required)
+			message_admins("Storyteller failed to pick an event for track of [track] due to insufficient population. (required: [pop_required] active pop for [track]. Current: [mode.active_players])")
```

Comment:
logging, see below

Reply:
success needs to be logged too- if storyteller does anything even remotely meaningful to the round it needs to be logged.